CONTRIBUTING.md

```diff
@@ -0,0 +1,80 @@
+code style, commit messages, testing setups, and more to help you get started.
+
+## Repository Structure
+
+The Hydroflow repo is set up as a monorepo and [Cargo workspace](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html).
+Relative to the repository root:
+
+* `hydroflow` is the main Hydroflow package, containing the Hydroflow runtime. It re-exports the
+  surface syntax macros which are in `hydroflow_macro` and `hydroflow_lang`. The runtime is the
+  "scheduled layer" while the surface syntax compiler is the "compiled layer".
+* `hydroflow_plus` and related packages contain Hydroflow+, which is a functional syntax built on
+  top of `hydroflow`.
+* `hydroflow_datalog` provides a datalog compiler, based on top of the Hydroflow surface syntax.
+* `docs` is the [Hydro.run](https://hydro.run/) website. `website_playground` contains the
+  playground portion of the website, used for compiling Hydroflow in-browser via WASM.
+* `benches` contains some microbenchmarks for Hydroflow and other frameworks.
+* `design_docs` contains old point-in-time design docs for Hydroflow's architecture.
+
+There are several subpackages included that are used by Hydroflow but are more general-purpose:
+
+* `stageleft` is a framework for staged programming in Rust, used by `hydroflow_plus`.
+* `lattices` is a abstract algebra library, originally for lattice types.
+* `variadics` is a crate for emulating variadic generics using tuple lists.
+* `pusherator` is a rudimentary library providing push-based iterators.
+* `multiplatform_test` provides a convenience macro for specifying and initializing tests on
+  various platforms.
+
+There are a few auxillary repositories as well:
+
+* [`hydro-project/hydroflow-template`](https://github.com/hydro-project/hydroflow-template)
+  provides a way to create a new Hydroflow user project, via `cargo generate hydro-project/hydroflow-template`.
+* [`hydro-project/rust-sitter`](https://github.com/hydro-project/rust-sitter) provides a
+  [Tree-sitter](https://tree-sitter.github.io/tree-sitter/)-based parser generator interface, used
+  by `hydroflow_datalog`.
+
+## Setup
+
+Some parts of the Hydroflow repo require a relatively recent version of Python 3, maybe 3.10 or
+later. On Mac, installing directly from python.org may work if `brew install` doesn't.
+
+## Commit Messages
+
+Commit messages should follow the [Conventional Commits specification](https://www.conventionalcommits.org/)
+when submitting a PR. Alternatively, GitHub's "Squash and merge" feature can be used to adjust the
+commit message, however those messages are not checked by CI.
+
+Commit messages are used to generate changelogs. See [Releasing](#releasing) for more.
+
+## Pull Requests
+
+CI runs a comprehensive set of tests on PRs before they are merged. This includes format and lint
+checks. To run (most of) the tests locally, you can run `./precheck.bash`. Note that this will
```

Comment:
Think precheck doesn't do WASM, correct? Any objections to adding the following to `precheck.bash`?

```shell
CARGO_TARGET_WASM32_UNKNOWN_UNKNOWN_RUNNER=wasm-bindgen-test-runner cargo test -p hydroflow --target wasm32-unknown-unknown
```